File: app/src/main/java/com/yibao/music/fragment/SongCategoryFragment.java
package com.yibao.music.fragment;

import android.os.Bundle;
import android.util.SparseBooleanArray;
import android.view.View;

import androidx.annotation.Nullable;

import com.yibao.music.R;
import com.yibao.music.adapter.SongAdapter;
import com.yibao.music.base.BaseLazyFragment;
import com.yibao.music.base.BaseRvAdapter;
import com.yibao.music.fragment.dialogfrag.MoreMenuBottomDialog;
import com.yibao.music.model.MusicBean;
import com.yibao.music.util.Constants;
import com.yibao.music.util.LogUtil;
import com.yibao.music.util.MusicListUtil;
import com.yibao.music.util.SnakbarUtil;
import com.yibao.music.view.music.MusicView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * @项目名： ArtisanMusic
 * @包名： com.yibao.music.artisanlist
 * @文件名: SongFragment
 * @author: lsp
 * @创建时间: 2018/2/4 21:45
 * @描述： {显示音乐分类列表}
 */

public class SongCategoryFragment extends BaseLazyFragment {

    @BindView(R.id.musci_view)
    MusicView mMusicView;
    private SongAdapter mSongAdapter;
    private int mPosition;
    private boolean isShowSlidBar = false;
    private boolean isItemSelectStatus = true;
    private static final String MUSIC_POSITION = "position";
    private SparseBooleanArray mSparseBooleanArray = new SparseBooleanArray();
    private List<MusicBean> mSelectList = new ArrayList<>();
    private Disposable mDeleteSongDisposable;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle arguments = getArguments();
        if (arguments != null) {
            mPosition = arguments.getInt(MUSIC_POSITION);
        }
    }

    @Override
    protected void initView(View view) {
        initData();
        initListener();
    }


    private void initListener() {
        mSongAdapter.setOnItemMenuListener(new BaseRvAdapter.OnOpenItemMoreMenuListener() {
            @Override
            public void openClickMoreMenu(int position, MusicBean musicBean) {
                LogUtil.d(TAG, "more menu click");
                MoreMenuBottomDialog.newInstance(musicBean, position, false, false).getBottomDialog(SongCategoryFragment.this.getActivity());
            }
        });
        mSongAdapter.setItemListener((bean, position, isEditStatus) -> {
            if (isEditStatus) {
                mSparseBooleanArray.put(position, true);
                updateSelected(bean);
                mSongAdapter.notifyDataSetChanged();
            }
        });
        mSongAdapter.setCheckBoxClickListener((bean, isChecked, position) -> {
            LogUtil.d(TAG, bean.getTitle() + " == " + isChecked);
            mSparseBooleanArray.put(position, isChecked);
            updateSelected(bean);
            mSongAdapter.notifyDataSetChanged();
        });
    }


    @Override
    protected int getContentViewId() {
        return R.layout.category_fragment;
    }

    @Override
    protected boolean getIsOpenDetail() {
        return isItemSelectStatus;
    }

    protected void initRxBusData() {
        disposeToolbar();
        if (mEditDisposable == null) {
            mEditDisposable = mBus.toObservableType(Constants.SONG_FAG_EDIT, Object.class).subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(o -> SongCategoryFragment.this.changeEditStatus((Integer) o));
        }
        if (mDeleteSongDisposable == null) {
            mDeleteSongDisposable = mBus.toObservableType(Constants.DELETE_SONG, Object.class)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(integer -> {
                        int position = (int) integer;
                        LogUtil.d(TAG, "收到删除    " + position);
                        mSongAdapter.deleteSong(position);
                    });

        }
    }

    @Override
    protected void deleteItem(int musicPosition) {
        super.deleteItem(musicPosition);
        mSongAdapter.notifyItemRemoved(musicPosition);
    }


    private void updateSelected(MusicBean bean) {
        if (mSelectList.contains(bean)) {
            mSelectList.remove(bean);
        } else {
            mSelectList.add(bean);
        }
    }

    protected void initData() {
        List<MusicBean> musicBeanList = mMusicBeanDao.queryBuilder().list();
        switch (mPosition) {
            case 0:
                List<MusicBean> abcList = MusicListUtil.sortMusicAbc(musicBeanList);
                setNotAllSelected(abcList);
                isShowSlidBar = true;
                mSongAdapter = new SongAdapter(mActivity, abcList, mSparseBooleanArray, Constants.NUMBER_ZERO, Constants.NUMBER_ZERO);
                break;
            case 1:
                List<MusicBean> scoreList = MusicListUtil.sortMusicList(musicBeanList, Constants.SORT_SCORE);
                setNotAllSelected(scoreList);
                mSongAdapter = new SongAdapter(mActivity, scoreList, mSparseBooleanArray, Constants.NUMBER_ONE, Constants.NUMBER_ONE);
                break;
            case 2:
                List<MusicBean> playFrequencyList = MusicListUtil.sortMusicList(musicBeanList, Constants.SORT_FREQUENCY);
                setNotAllSelected(playFrequencyList);
                mSongAdapter = new SongAdapter(mActivity, playFrequencyList, mSparseBooleanArray, Constants.NUMBER_ONE, Constants.NUMBER_TWO);
                break;
            case 3:
                List<MusicBean> addTimeList = MusicListUtil.sortMusicList(musicBeanList, Constants.SORT_DOWN_TIME);
                setNotAllSelected(addTimeList);
                mSongAdapter = new SongAdapter(mActivity, addTimeList, mSparseBooleanArray, Constants.NUMBER_ONE, Constants.NUMBER_ZERO);
                break;
            default:
                break;
        }
        mMusicView.setAdapter(mActivity, Constants.NUMBER_ONE, isShowSlidBar, mSongAdapter);
    }

    private void setNotAllSelected(List<MusicBean> listBeanList) {
        for (int i = 0; i < listBeanList.size(); i++) {
            mSparseBooleanArray.put(i, false);
        }
    }

    private void changeEditStatus(int currentIndex) {
        if (currentIndex == Constants.NUMBER_ONE) {
            closeEditStatus();
        } else if (currentIndex == Constants.NUMBER_TWO) {
            // 删除已选择的条目
            deleteListItem();
        }
    }

    private void deleteListItem() {
        LogUtil.d(TAG, "Size " + mSelectList.size());
        if (mSelectList.size() > Constants.NUMBER_ZERO) {
            for (MusicBean musicBean : mSelectList) {
                LogUtil.d(TAG, musicBean.getTitle());
//                FileUtil.deleteFile(new File(musicBean.getSongUrl()));
//                mMusicBeanDao.delete(musicBean);
            }
            mSongAdapter.setItemSelectStatus(false);
            mSongAdapter.setNewData(getSongList());
//            mBus.post(Constants.FRAGMENT_SONG, Constants.NUMBER_ZERO);
        } else {
            SnakbarUtil.favoriteSuccessView(mMusicView, "没有选中条目");
        }
    }

    private List<MusicBean> getSongList() {

        List<MusicBean> musicBeans = mMusicBeanDao.queryBuilder().list();
        return MusicListUtil.sortMusicAbc(musicBeans);
    }


    private void closeEditStatus() {
        interceptBackEvent(isItemSelectStatus ? Constants.NUMBER_ELEVEN : Constants.NUMBER_ZERO);
        mSongAdapter.setItemSelectStatus(isItemSelectStatus);
        isItemSelectStatus = !isItemSelectStatus;
        mSparseBooleanArray.clear();
        if (mSelectList.size() > 0) {
            mSelectList.clear();
        }
    }

    @Override
    protected void handleDetailsBack(int detailFlag) {
        if (detailFlag == Constants.NUMBER_ELEVEN) {
            mSongAdapter.setItemSelectStatus(false);
            mBus.post(Constants.FRAGMENT_SONG, Constants.NUMBER_ZERO);
            isItemSelectStatus = !isItemSelectStatus;
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (mDeleteSongDisposable != null) {
            mDeleteSongDisposable.dispose();
            mDeleteSongDisposable = null;

        }
    }

    public static SongCategoryFragment newInstance(int position) {
        Bundle args = new Bundle();
        SongCategoryFragment fragment = new SongCategoryFragment();
        args.putInt(MUSIC_POSITION, position);
        fragment.setArguments(args);
        return fragment;
    }

}

File: app/src/main/java/com/yibao/music/fragment/SongCategoryFragment.kt
package com.yibao.music.fragment

import android.content.Intent
import android.os.Bundle
import android.util.SparseBooleanArray
import android.view.View
import com.yibao.music.R
import com.yibao.music.activity.PlayListActivity
import com.yibao.music.adapter.SongAdapter
import com.yibao.music.base.bindings.BaseBindingAdapter
import com.yibao.music.base.bindings.BaseMusicFragmentDev
import com.yibao.music.databinding.CategoryFragmentBinding
import com.yibao.music.fragment.dialogfrag.MoreMenuBottomDialog
import com.yibao.music.model.MusicBean
import com.yibao.music.util.Constant
import com.yibao.music.util.LogUtil
import com.yibao.music.util.ToastUtil
import com.yibao.music.viewmodel.SongViewModel

/**
 * @项目名： ArtisanMusic
 * @包名： com.yibao.music.artisanlist
 * @文件名: SongFragment
 * @author: lsp
 * @创建时间: 2018/2/4 21:45
 * @描述： {显示音乐分类列表}
 */
class SongCategoryFragment : BaseMusicFragmentDev<CategoryFragmentBinding>(), View.OnClickListener {

    private val mViewModel: SongViewModel by lazy { gets(SongViewModel::class.java) }
    private var mStateArray = SparseBooleanArray()
    private val mSelectList = ArrayList<MusicBean>()


    private var mList = ArrayList<MusicBean>()
    override fun initView() {
        mBinding.ivSongAddToList.setOnClickListener(this)
        mBinding.ivSongAddToPlay.setOnClickListener(this)
    }

    override fun initData() {

    }

    override fun onResume() {
        super.onResume()
        val position = requireArguments().getInt(Constant.POSITION)
        mViewModel.getMusicList(position)
        mViewModel.listModel.observe(this) { musicList ->
            if (musicList.isNotEmpty()) {
                mList.clear()
                mList.addAll(musicList)
                initAdapter(musicList, position)

            }

        }

    }

    private fun initAdapter(musicList: List<MusicBean>, position: Int) {
        when (position) {
            0 -> {
                val adapter = SongAdapter(
                    mActivity, musicList, mStateArray, true, 0, 1
                )
                setData(adapter)
            }

            1 -> {

                val adapter = SongAdapter(
                    mActivity, musicList, mStateArray, false, 1, 2
                )


                setData(adapter)
            }

            2 -> {
                val adapter = SongAdapter(
                    mActivity, musicList, mStateArray, false, 2, 3
                )

                setData(adapter)
            }

            3 -> {
                val adapter = SongAdapter(
                    mActivity, musicList, mStateArray, false, 0, 4
                )

                setData(adapter)
            }
        }


    }


    private fun setData(adapter: SongAdapter) {
        mBinding.musicView.setAdapter(requireActivity(), Constant.NUMBER_ONE, true, adapter)
        adapter.setOnItemMenuListener(object : BaseBindingAdapter.OnOpenItemMoreMenuListener {
            override fun openClickMoreMenu(position: Int, musicBean: MusicBean) {
                MoreMenuBottomDialog.newInstance(
                    musicBean,
                    position,
                    isNeedScore = false,
                    isNeedSetTime = false
                ).getBottomDialog(requireActivity())
            }
        })

        adapter.setCheckBoxClickListener(object :
            BaseBindingAdapter.OnCheckBoxClickListener<MusicBean> {
            override fun checkboxChange(t: MusicBean, isChecked: Boolean, position: Int) {
                LogUtil.d(mTag, t.title + " == " + isChecked)
                mStateArray.put(position, isChecked)
                updateSelected(t)
                adapter.notifyDataSetChanged()
            }
        })

        adapter.setItemLongClickListener(object :
            BaseBindingAdapter.ItemLongClickListener<MusicBean> {
            override fun longClickItem(musicInfo: MusicBean, currentPosition: Int) {
                LogUtil.d(mTag, "长按了")
            }
        })

        adapter.setCbShowListener(object : SongAdapter.OnShowCbListener {
            override fun showCb(showCb: Boolean) {
                LogUtil.d(mTag, "IS SHOW CB  $showCb")
                isShowCb = showCb
                if (!showCb) {
                    mSelectList.clear()
                    setNotAllSelected(mList)
                }
                updateAddToListBtn(showCb)
            }
        })
    }

    private var isShowCb = false
    override fun deleteItem(musicPosition: Int) {
        super.deleteItem(musicPosition)
//        mSongAdapter.notifyItemRemoved(musicPosition)
    }

    private fun updateSelected(bean: MusicBean) {
        if (mSelectList.contains(bean)) {
            mSelectList.remove(bean)
        } else {
            mSelectList.add(bean)
        }
        LogUtil.d(mTag, "选中长度：  ${mSelectList.size}")
    }

    private fun setNotAllSelected(listBeanList: List<MusicBean>) {
        for (i in listBeanList.indices) {
            mStateArray.put(i, false)
        }
    }


    companion object {

        @JvmStatic
        fun newInstance(position: Int): SongCategoryFragment {

            val args = Bundle()
            val fragment = SongCategoryFragment()
            args.putInt(Constant.POSITION, position)
            fragment.arguments = args
            return fragment
        }
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.iv_song_add_to_list -> {
                LogUtil.d(mTag, "添加长度：  ${mSelectList.size}")
                startPlayListActivity()
            }

            R.id.iv_song_add_to_play -> {
                LogUtil.d(mTag, "添加长度：  ${mSelectList.size}")
            }
        }

    }

    private fun startPlayListActivity() {
        if (mSelectList.isEmpty()) {
            ToastUtil.show(requireActivity(), "请选择要添加的歌曲")
        } else {
            val arrayList = ArrayList<String>()
            for (musicBean in mSelectList) {
                arrayList.add(musicBean.title)
            }
            val intent = Intent(context, PlayListActivity::class.java)
            intent.putStringArrayListExtra(Constant.ADD_TO_LIST, arrayList)
            startActivity(intent)
        }
    }

    private fun updateAddToListBtn(showCb: Boolean) {
        mBinding.groupSongAdd.visibility = if (showCb) View.VISIBLE else View.GONE
    }

    /**
     * 列表进入选中状态时，需要处理返回事件。
     */
    override fun onBackPressed(): Boolean {
        updateAddToListBtn(!isShowCb)
        if (isShowCb) {
            // 选中状态，按返回键取消选中状态
            mBinding.groupSongAdd
            // 隐藏添加到插入列表按钮。
            mBinding.musicView.updateCbState()
            isShowCb = false
            return true
        } else {
            return false
        }

    }
}